src/server/zset_family.cc

```diff
@@ -2840,54 +2848,108 @@ void SortIfNeeded(GeoArray* ga, Sorting sorting, uint64_t count) {
+    }
+    if (write_mode) {
+      cntx->transaction->Conclude();
+    }
     return;
   }
 
   auto* rb = static_cast<RedisReplyBuilder*>(cntx->reply_builder());
   // if no matching results, the user gets an empty reply.
   if (ga.empty()) {
     rb->SendNull();
+    if (write_mode) {
+      cntx->transaction->Conclude();
+    }
     return;
   }
 
   // sort and trim by count
   SortIfNeeded(&ga, geo_ops.sorting, geo_ops.count);
 
-  // generate reply array withdist, withcoords, withhash
-  int record_size = 1;
-  if (geo_ops.withdist) {
-    record_size++;
-  }
-  if (geo_ops.withhash) {
-    record_size++;
-  }
-  if (geo_ops.withcoord) {
-    record_size++;
-  }
-  rb->StartArray(ga.size());
-  for (const auto& p : ga) {
-    // [member, dist, x, y, hash]
-    rb->StartArray(record_size);
-    rb->SendBulkString(p.member);
+  if (geo_ops.store == GeoStoreType::kNoStore) {
+    // case 1: read mode
+    // case 2: write mode, kNoStore
+    //  generate reply array withdist, withcoords, withhash
+    int record_size = 1;
     if (geo_ops.withdist) {
-      rb->SendDouble(p.dist / geo_ops.conversion);
+      record_size++;
     }
     if (geo_ops.withhash) {
-      rb->SendDouble(p.score);
+      record_size++;
     }
     if (geo_ops.withcoord) {
-      rb->StartArray(2);
-      rb->SendDouble(p.longitude);
-      rb->SendDouble(p.latitude);
+      record_size++;
+    }
+    rb->StartArray(ga.size());
+    for (const auto& p : ga) {
+      // [member, dist, x, y, hash]
+      rb->StartArray(record_size);
+      rb->SendBulkString(p.member);
+      if (geo_ops.withdist) {
+        rb->SendDouble(p.dist / geo_ops.conversion);
+      }
+      if (geo_ops.withhash) {
+        rb->SendDouble(p.score);
+      }
+      if (geo_ops.withcoord) {
+        rb->StartArray(2);
+        rb->SendDouble(p.longitude);
+        rb->SendDouble(p.latitude);
+      }
     }
+  } else {
+    // case 3: write mode, !kNoStore
+    DCHECK(geo_ops.store == GeoStoreType::kStoreDist || geo_ops.store == GeoStoreType::kStoreHash);
+    ShardId dest_shard = Shard(store_key, ga.size());
+    DVLOG(1) << "store shard:" << dest_shard << ", key " << store_key;
+    AddResult add_result;
+    vector<ScoredMemberView> smvec;
+    for (const auto& p : ga) {
+      if (geo_ops.store == GeoStoreType::kStoreDist) {
+        smvec.emplace_back(p.dist / geo_ops.conversion, p.member);
+      } else {
+        DCHECK(geo_ops.store == GeoStoreType::kStoreHash);
+        smvec.emplace_back(p.score, p.member);
+      }
+    }
+
+    auto store_cb = [&](Transaction* t, EngineShard* shard) {
+      if (shard->shard_id() == dest_shard) {
+        ZParams zparams;
+        zparams.override = true;
+        add_result =
+            OpAdd(t->GetOpArgs(shard), zparams, store_key, ScoredMemberSpan{smvec}).value();
+      }
+      return OpStatus::OK;
+    };
+    cntx->transaction->Execute(std::move(store_cb), false);
+
+    rb->SendLong(smvec.size());
+  }
+  if (write_mode) {
```

Comment:
I do not think you need this conclude step in the normal flow that does not expect errors. You always know what's your last hop is and that should have `Execute(..., true)`

Reply:
The last hop could vary. When STORE is not used, the last hop would be hop 2.
If I can't use conclude in this way, I'll have to add a argument "do_store" for the whole workflow.

Reply:
Also, after hop 2, every conclude() would have to check do_store. If already called conclude after hop 2, skip and return.

Reply:
There are at most 3 hops
hop1. Getting latitude longitude from member name
hop2. Getting nearby nodes around <latitude ,longitude >
hop3. write to store key if needed

Reply:
Conclude is also a hop. Seem my previous comment  at line 2947